### DISCUSSION TITLE
Algorithms proposal

### DISCUSSION THREAD

**Discussion by kpetridis24** [networkx/networkx#5448] [Algorithms]
There is no class to store large sparse graphs in a format such that it can be handled by the memory (if i am mistaken please let me know). So i plan on creating a class SparseGraph that's gonna keep the CSR format (using SciPy) for a graph, and include functions for high performance triangle counting, maybe k-NN, etc.

**What's your opinion?**

**Comment by MridulS** [comment]
Thanks for bringing this up @kpetridis24! This is something that has been brought up multiple times over time, you can go through some old discussion https://github.com/networkx/networkx/issues/4135 and we are actively looking into implementing some kind of support for `SparseGraph` in networkx. We don't have a concrete proposal yet!

Recently we have started looking into GraphBLAS (the python interface https://github.com/metagraph-dev/grblas) which will let us take advantage of more performant sparse graph algorithms and directly dispatch the computation from NetworkX to a sparse implementation.

Outside of NetworkX there is also [scipy.sparse.csgraph](https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/reference/sparse.csgraph.html#module-scipy.sparse.csgraph) which implements a subset of sparse graph algorithms, wrapping them up and implementing them in NetworkX would be another thing that we can look into, assuming there is a significant performance boost including the overhead of NX glue code.

**Reply by kpetridis24** [reply]
I would be very interested in putting time and effort to this. I have already started planning about how this is going to be implemented. I have experience with working with sparse graphs and BLAS libraries. We could make a new class as we said, and implement a ton of high performance functions using BLAS, scipy and other high performance techniques. Do you think this is going to be a good project for 2022 GSoC. If yes, i would be extremely interested in taking this on.
